Conda package updates to support newer Xcode versions and M1 Apple Silicon

cticenhour opened the pull request:
Updates Mac packages to Clang 12.0.1 and adds support for Xcode 12.5.x through 13.1 (packaged with macOS Monterey and available in Big Sur) 

Adds experimental support for M1 Apple Silicon Macs <ins>**for developer use only.**</ins> MOOSE and modules will build, but still has a handful of testing failures. More work is needed here. 

Closes #16291
Refs #18954 

Confirmed working (`test/run_tests`) on Intel-based Mac:
| OS | CLT | XCode | `moose-libmesh` | `moose-petsc` | `moose-mpich` |
| - | - | - | - | - | - |
| 12.0.1 | 13.1 | | ✅ | | ✅  |
| 11.6 | 12.4 | | ✅  | | ✅ |
| 11.6.1 | | 13.1 | ✅ | ✅ | ✅ |
| 11.2.3 | | 12.4 | ✅  | | ✅  |
| 10.15.7 | | 11.5 | ✅  | ✅  | |

Commit:
Update to clang 12.0.1 and have semi-working ARM64 stack (#16291)

Includes squashed commits:
- Update to clang 12.0.1 (#16291)
- Changes that seem to work on linux
- Mac adjustments
- Fixup moose-compilers
- Changes necessary for semi-functional ARM64 MPICH.
- Culmination of changes to produce a viable ARM64 stack.
- Include PETSc patch necessary for ARM, build superlu_dist
- Changes necessary for Linux
- Modify jinja template for moose-tools compilers

cticenhour committed:
Adjustments that build with Intel Xcode 13 (#16291)

Includes squashed commits:
- Fix mpich build config typo for osx-64
- Update osx-64 versions and builds for Intel Xcode 13
- Add libclang to mpich and change a few package names to make sure we include all dependencies compared to current Intel build.

Commit:
Changes that represent a working conda stack for M1, Linux, and the older Intel Mac arch (#16291)

Includes squashed commits/changes:
- Clang 12 and Python >=3.8 is missing dependencies for PyQT
- Add sympy
- Add libclang
- Downgrade cctools and ld so that the packages match cross-platform. There
appears to be a bug (maybe its not a bug?) when performing `conda index`,
one arch seems to override the other with the dependency chain.
- Adjusting versions for all platforms. The following are updates available
for Arm64 and Linux arches.
- A working M1, Linux, Intel older Xcode build.sh

cticenhour committed:
A confirmed working Intel build.sh for 12.4, 13, and 13.1 (#16291)

Also don't unset LDFLAGS before arch determination

Commit:
Clean up  mpich/build.sh (#16291)

Some changes which seem to work for my three setups.

cticenhour committed:
Remove moose-compilers and attempt to fixup moose-pyhit (#16291)

Commit:
More awesome changes

Commit:
More dependency reduction

Commit:
Refactor naming schemes

Use <package version build_version> throughout conda_build_config.yaml (where applicable)

cticenhour committed:
Add openmp flag to PETSc update script

This fixes the PETSc build when using only the new moose-mpich package, 
where MUMPS was failing due to missing OpenMP symbols.

cticenhour committed:
Fixup chemical_reactions aqueous equilibirum test for threads and parallel.

After tightening both absolute and relative nonlinear tolerances, one 
variable was *just* out of spec in only one timestep in the middle of 
the simulation. Re-golded after inspection. Now test runs successfully 
with both threads and parallel, so removing test spec restriction is 
justified.

cticenhour committed:
Add deepdiff to moose-tools package

cticenhour committed:
Fixup phase_field tests by tightening tolerances

dschwen commented on conda/mpich/meta.yaml:
```diff
@@ -20,47 +20,48 @@ build:
+
+  test:
+    commands:
+      - echo "It works!"
```
flawless test!

cticenhour replied:
Hahaha - we plan to restore it. I believe it was getting in the way of Jason building packages on M1.

cticenhour replied:
Restored in 39ffb00 and will fixup with a rebase before merge.

fdkong commented on conda/petsc/build.sh:
```diff
@@ -44,29 +47,32 @@ BUILD_CONFIG=`cat <<"EOF"
   --download-parmetis=1 \
+  --download-strumpack=1 \
   --download-scalapack=1 \
+  --download-fblaslapack=1 \
```
Should we just call "update_and_rebuild_petsc.sh" from here?

loganharbour replied:
If possible - I would really prefer this. It would be nice if we could add the additional conda require things to `update_and_rebuild_petsc.sh` so that we can for sure keep them in sync as much as possible

cticenhour replied:
It's a good idea. I'll look into it!

cticenhour replied:
@fdkong, I think we've addressed this with the combo of a5e3b39 and 334ca0c if you want to take a look.

fdkong commented on conda/petsc/arm64_framework.patch:
```diff
@@ -0,0 +1,12 @@
+       lines = [s for s in lines if s.find(' was built for newer macOS version') < 0]
+       lines = [s for s in lines if s.find(' was built for newer OSX version') < 0]
+       lines = [s for s in lines if s.find(' stack subq instruction is too different from dwarf stack size') < 0]
++      lines = [s for s in lines if s.find(' could not create compact unwind for') < 0]
```
I will check if the new release has this fix. We do not want to do this kind of thing too often . It will be hard to maintain multiple versions of petsc.

cticenhour replied:
@fdkong did you look into a newer release? @milljm Do you have any comment on this ARM64 PETSc patch to address Fande's concern?

cticenhour replied:
@fdkong I would like to work on getting this merged into the framework today or tomorrow. Since ARM64 on MacOS is experimental for the moment, we will have a few more PRs coming to address M1 compatibility and hopefully your concerns about this patch. Do you have any other review comments to share in this PR?

fdkong replied:
We can go this patch. I will patch future petsc

cticenhour replied:
Keep in mind that this is not in production. A package for arm64 is not currently being built. We might need to make changes (or remove this patch altogether) as we develop more for this platform.

cticenhour replied:
Very much a work in progress

cticenhour committed:
Fixup libmesh package build number in config after rebase

loganharbour committed:
Separate petsc configure into another script for use in all petsc builds

refs #16291

cticenhour committed:
Make sure conda build moves configure_petsc.sh into working directory

Also clean up conda build script and fixup headerpad-max-install-names 
flag.

loganharbour commented:
Single failure with `moose-mpich moose-tools` and self compiled libmesh/petsc on Monterey with current CLT and Big Sur with 12.4 CLT:

```
combined/test:break_mesh_interface_contact.break_mesh_interface_contact: Sideset Distribution Factors:
combined/test:break_mesh_interface_contact.break_mesh_interface_contact:   --------- Time step 1, 0.0000000e+00 ~ 0.0000000e+00, rel diff:  0.00000e+00 ---------
combined/test:break_mesh_interface_contact.break_mesh_interface_contact: Nodal variables:
combined/test:break_mesh_interface_contact.break_mesh_interface_contact: Element variables:
combined/test:break_mesh_interface_contact.break_mesh_interface_contact:   --------- Time step 2, 1.0000000e-02 ~ 1.0000000e-02, rel diff:  0.00000e+00 ---------
combined/test:break_mesh_interface_contact.break_mesh_interface_contact: Nodal variables:
combined/test:break_mesh_interface_contact.break_mesh_interface_contact: Element variables:
combined/test:break_mesh_interface_contact.break_mesh_interface_contact:    stress_yy           rel diff: -1.1168611e-02 ~ -1.1168734e-02 = 1.09869e-05 (block 2, elmt 18)
combined/test:break_mesh_interface_contact.break_mesh_interface_contact:   --------- Time step 3, 2.0000000e-02 ~ 2.0000000e-02, rel diff:  0.00000e+00 ---------
```

moosebuild commented:
Job [Test timings](https://civet.inl.gov/job/869626/) on 334ca0c wanted to post the following:

View timings [here](https://mooseframework.inl.gov/docs/PRs/19230/timing)

This comment will be updated on new commits.

cticenhour committed:
Add Casey to CODEOWNERS for conda (#16291)

cticenhour committed:
Separate libmesh configure into another script for use in all libmesh builds (#16291)

cticenhour commented:
Note to all: I've broken the heck out of the testing due to the libmesh configure changes - working on it.

cticenhour commented on scripts/configure_libmesh.sh:
```diff
@@ -0,0 +1,53 @@
+    exit 1
+  fi
+
+  if [[ -n "$INSTALL_BINARY" ]]; then
+    echo "INFO: INSTALL_BINARY set"
+  else
+    export INSTALL_BINARY="${SCRIPT_DIR}/../libmesh/build-aux/install-sh -C"
+  fi
```
@loganharbour this bit could remain in `update_and_rebuild_libmesh.sh`, but we then would need to make sure that `INSTALL_BINARY` is set in `build.sh` as well. If I leave this here, it might be worth adding a comment in the old location so that this can be easily found again by someone looking at the update script. Thoughts?

cticenhour replied:
Comment added in 7763178

cticenhour commented:
Cataloging the netCDF configuration error seen during the troubleshooting of 8f2df4c, which necessitated the move of INSTALL_BINARY to `configure_libmesh.sh`. The error can be seen by clicking the dropdown below:

<details closed>
<summary>Click to open</summary>

```
=== configuring in contrib/netcdf/v4 (/Users/icenct/projects/moose/libmesh/contrib/netcdf/v4)
configure: running /bin/sh ./configure --disable-option-checking '--prefix=/Users/icenct/projects/moose/scripts/../libmesh/installed'  '--enable-silent-rules' '--enable-unique-id' '--disable-warnings' '--enable-glibcxx-debugging' '--with-thread-model=openmp' '--disable-maintainer-mode' '--enable-petsc-hypre-required' '--enable-metaphysicl-required' '--with-cxx-std-min=2014' '--with-methods=opt devel dbg' 'INSTALL=/Users/icenct/projects/moose/scripts/../libmesh/build-aux/install-sh' 'build_alias=x86_64-apple-darwin13.4.0' 'host_alias=x86_64-apple-darwin13.4.0' 'PETSC_DIR=/Users/icenct/miniconda3/envs/testing-libmesh-opt' 'CC=mpicc' 'CFLAGS=-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include' 'LDFLAGS=-Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/icenct/miniconda3/envs/testing-libmesh-opt/lib -L/Users/icenct/miniconda3/envs/testing-libmesh-opt/lib' 'CPPFLAGS=-D_FORTIFY_SOURCE=2 -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include' 'FC=mpif90' 'F77=mpif77' 'FFLAGS=-march=core2 -mtune=haswell -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include' 'CXX=mpicxx' 'CXXFLAGS=-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong  -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include' 'METHODS=opt devel dbg' '--disable-netcdf-4' '--disable-dap' '--disable-curl' '--disable-testsets' 'CXX=mpicxx' 'CC=mpicc' 'F77=mpif77' 'FC=mpif90' 'CPPFLAGS= -D_FORTIFY_SOURCE=2 -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include' 'LIBS= ' --cache-file=../../../config.cache --srcdir=.
configure: loading cache ../../../config.cache
configure: error: `LIBS' was not set in the previous run
configure: warning: ignoring whitespace changes in `CPPFLAGS' since the previous run:
configure:   former value:  `-D_FORTIFY_SOURCE=2 -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include'
configure:   current value: ` -D_FORTIFY_SOURCE=2 -isystem /Users/icenct/miniconda3/envs/testing-libmesh-opt/include'
configure: error: in `/Users/icenct/projects/moose/libmesh/contrib/netcdf/v4':
configure: error: changes in the environment can compromise the build
configure: error: run `make distclean' and/or `rm ../../../config.cache' and start over
configure: error: ./configure failed for contrib/netcdf/v4
Running make -j 56...
make: *** No targets specified and no makefile found.  Stop.
```

</details>

I am adding this here for a log in case an issue related to this comes up in the future.

cticenhour committed:
Fixup netCDF configure error (#19230)

cticenhour committed:
Fixup conda (#16291)

Conda build considers `INSTALL_BINARY` an "unbound variable" without 
setting it in build.sh.

cticenhour commented on conda/petsc/build.sh:
```diff
@@ -24,7 +24,7 @@ if [[ $(uname) == Darwin ]]; then
         FTUNING="-I$PREFIX/include"
     fi
-    LDFLAGS="${LDFLAGS:-} -Wl,-headerpad_max_install_names"
+    LDFLAGS="${LDFLAGS:-} -Wl,-headerpad-max-install-names"
```
@milljm @dschwen Looking at the M1 PETSc configure error you posted reminded me of this needed change on my Intel Mac, which threw a similar error. Not sure what caused it yet (I assumed originally we were just using an old form of the flag), but I wanted to make note.

loganharbour committed:
Loosen tolerance on test

refs #16291

cticenhour committed:
Actually add deepdiff to moose-tools package (#16291)

tophmatthews commented:
I'm running a local workshop on moose here this week and next, with users that are stuck with Big Sur. Is there a way to fix the recent xcode updates before the M1 is working?

cticenhour commented:
@tophmatthews Yes, this is ready to merge with Intel. I'm confirming M1 is in an experimental state before pressing GO, but I hope to get this merged today or tomorrow.

cticenhour commented:
If you don't mind your users running on our PR channel, here are the instructions for that: 

```
conda deactivate
conda install mamba 
conda create -y -n moose-libmesh-test
conda activate moose-libmesh-test
mamba install -c https://conda.software.inl.gov/moose/pr-19230 moose-libmesh moose-tools 
```

Mamba is much faster than conda at doing installs, but that part is optional.

tophmatthews commented:
ya...working with newbies, so I don't want to mess to much with their systems...

cticenhour commented:
I will ping you when this makes it to master and can be used as normal!

cticenhour committed:
Add M1 Monterey conda build instructions (#18954)

milljm reviewed (dismissed):
Thanks for taking the time to create this!

loganharbour reviewed (approved):
Approving content that is not my own, @friedmud should do a review for all content

friedmud reviewed (approved):
@loganharbour told me to approve this.

cticenhour commented:
Still need buy-in from @fdkong

moosebuild commented:
Job [Apple Silicon M1](https://civet.inl.gov/job/873157/) on e2aaeb0 : canceled by @cticenhour

Apple Silicon M1 is a WIP recipe. Does not need to pass yet for successful PR merge.

tophmatthews commented:
you da :bomb: @cticenhour ! Thats for the hustle!
